0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,283 @@
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
+
+Third we dealt with the problem that there is not always an `address_hint` (i.e. Bob does not need to know the `address_hint` of Alice).
+We added the `address_hint` extension table that relates to `swap` on the `counter_party_peer_id` field (rather than on the `swap_local_id`).
+
+Note that the role could be factored out of the table, but we decided against it because it adds unnecessary complexity for now.
+
+What remains is a simple table for `swap` which contains only the information which all swaps have in common: `local_swap_id` `role` and `counter_party_peer_id`.
+
+=== Protocols
+
+[[HanProtocols]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/86bfc97bcea91602eaf3e40e6da7d71beaff9911/assets/db_schema/han_bitcoin_swap_protocols_without_state.puml&fmt=svg[HanBitcoinProtocols]
+
+In order to store details for each swap such as data we need for creating the HTLCs for each protocol we introduce `\{protocol\}_swap_details`, where `\{protocol\}` can be `han_bitcoin`, ``han_ethereum`, `herc20` or `halight` for now.
+Similar to the `Swaps` table, we add information which is only known at a later point of time into a sub-table.
+This resulted in quite an excess of sub-tables such as `han_bitcoin/ethereum_redeem_identity`, `han_bitcoin/ethereum_refund_identity`, etc.
+
+Some notable sub-tables:
+
+* `han_bitcoin/ethereum_swap_detail` contains data needed to create the HTLCs. We decided to give this table a generated primary key, rather than using `swap_local_id` as the primary key to allow same chain swaps.
+* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
+
+HErc20 and HaLight follow the same principles as described above.
+For completeness we depict `herc20` and `halight` diagrams below.
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/86bfc97bcea91602eaf3e40e6da7d71beaff9911/assets/db_schema/han_ethereum_swap_protocols_without_state.puml&fmt=svg[HanEthereumProtocols]
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/86bfc97bcea91602eaf3e40e6da7d71beaff9911/assets/db_schema/herc20_swap_protocols_without_state.puml&fmt=svg[Herc20Protocols]
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/86bfc97bcea91602eaf3e40e6da7d71beaff9911/assets/db_schema/halight_swap_protocols_without_state.puml&fmt=svg[HalightProtocols]
+
+
+==== Protocols with states:
```

Comment:
Is your intention to actually implement these tables also right away or are they just here for inspiration once we want to have persistent state? Can we, for the sake of moving forward, exclude these for now? :)

I wouldn't want to discuss the details now unless you actually want to include those right away as-well.

Reply:
I didn't want to throw away the work, so they are just here for inspiration :)